mudata/_core/mudata.py

```diff
@@ -404,137 +425,229 @@ def _update_attr(self, attr: str, axis: int, join_common: bool = False):
-                join="outer",
-                axis=axis,
-                sort=False,
-            )
+            # this occurs when join_common=True and we already have a global data frame, e.g. after reading from HDF5
+            if join_common:
+                sharedcols = data_mod.columns.intersection(data_global.columns)
+                data_global.rename(
+                    columns={col: f"global:{col}" for col in sharedcols}, inplace=True
+                )
+
+            # TODO: is using an attrmap here has any benefit when no duplicates or intersections?
 
-            data_common = pd.concat(
-                [_make_index_unique(getattr(a, attr)[columns_common]) for m, a in self.mod.items()],
-                join="outer",
-                axis=0,
-                sort=False,
+            if len(data_global) > 0:
+                data_mod = data_mod.join(data_global, how="left", sort=False)
+            data_mod.reset_index(level=list(range(1, data_mod.index.nlevels)), inplace=True)
+            data_mod.index.set_names(None, inplace=True)
+
+            if join_common:
+                for col in sharedcols:
+                    gcol = f"global:{col}"
+                    if np.array_equal(data_mod[col], data_mod[gcol]):
+                        data_mod.drop(columns=gcol, inplace=True)
+                    else:
+                        warnings.warn(
+                            f"Column {col} was present in {attr} but is also a common column in all modalities, and their contents differ. {attr}.{col} was renamed to {attr}.{gcol}."
+                        )
+
+            # Add data from global .obs/.var columns
+            # This might reduce the size of .obs/.var if observations/variables were removed
+            setattr(
+                # Original index is present in data_global
+                self,
+                "_" + attr,
+                data_mod,
             )
-            data_mod = data_mod.join(data_common, how="left", sort=False)
+
+            # Make simple attrmaps with anndata positions
+            # TODO: any edgecases?
+            mdict = dict()
+            incr = 0
+            for m in self.mod.keys():
+                mdict[m] = np.zeros(data_mod.shape[0])
+                mod_size = getattr(self.mod[m], attr).shape[0]
+                mdict[m][incr : (incr + mod_size)] = np.arange(mod_size) + 1
+                incr += mod_size
+
+        #
+        # General case: with duplicates and/or intersections
+        #
         else:
-            dfs = [
-                _make_index_unique(
-                    getattr(a, attr)
-                    .assign(**{rowcol: np.arange(getattr(a, attr).shape[0])})
-                    .add_prefix(m + ":")
+            if join_common:
+                # If all modalities have a column with the same name, it is not global
+                columns_common = reduce(
+                    np.intersect1d, [getattr(self.mod[mod], attr).columns for mod in self.mod]
```

Comment:
should we also switch to Pandas Index intersection here? Also, I think that piece of code is duplicated in the two branches (no duplicates/intersections and duplicates/intersections), it can probably be moved to above the `if not attr_intersecting and not attr_duplicated:`